feat: revamp timeouts

serramatutu opened the pull request:
This PR adds a new `timeout` parameter to the global `SemanticLayerClient` that gets passed down to the `GraphQLClient`.

The `timeout` parameter is a `TimeoutOptions` enum, which contains:
- `connection_timeout`: timeout to establish a connection to the servers
- `execute_timeout`: timeout to run an operation against the server (not counting retries)
- `tls_close_timeout`: timeout to close a TLS connection
- `total_timeout`: timeout for total wall time of an operation like `query()` or `dimensions()`, including retries, from the caller's point of view

Users can also pass in `int` or `float` values, which should be interpreted as equivalent to setting `total_timeout`. In practice, this gets converted to `TimeoutOptions` by assigning that same value to everything. Since everything should be upper-bounded by `total_timeout`, this should be fine.

Since I was at it, I deprecated the `timeout_ms` parameter in `ExponentialBackoff`. This value will be considered equivalent to `total_timeout` if provided, respecting the current behavior. If users use it, it will show a warning.

Finally, I improved the error handling logic of GraphQL clients so that internal `aiohttp` and `requests` timeout errors get re-raised as `dbtsl.TimeoutError`. This should make it easier for users to catch these errors without worrying about the underlying transport implementation.

You can review by commit.

aiguofer commented on dbtsl/api/graphql/client/asyncio.py:
```diff
@@ -31,6 +31,7 @@ def __init__(
         environment_id: int,
         auth_token: str,
         url_format: Optional[str] = None,
+        timeout_ms: Optional[int] = None,
```
I think we should define timeouts in seconds, not ms. There's no real reason to ever have a timeout be less than 1s.

serramatutu commented:
@aiguofer Yea, that's a good point. We also have an extra "timeout" which is the maximum time that can pass during backoff. I also saw that GQL has a TLS close timeout. So, in summary
- connect timeout
- execute timeout
- TLS close timeout
- backoff timeout (in-between requests)

Though, from looking at GQL code, it looks like it doesn't always use `execute_timeout` (only seems like it's used for async transports).

aiguofer commented:
> Though, from looking at GQL code, it looks like it doesn't always use execute_timeout (only seems like it's used for async transports).

Yeah that sounds familiar... lol I remember thinking it was pretty odd the way it was designed when I dug into the code

serramatutu commented:
Now that I think about it better, I think it's less confusing for users if we create a `TimeoutOptions` dataclass with all these options, and make a global `timeout` parameter which can be `TimeoutOptions | int | None`. This way if users don't wanna dig into all these params they can just slap an integer there that will get used for everything.

aiguofer commented:
I like the simplicity of that option. My only concern would be that a connection timeout and backoff timeout should be pretty short (say 5s or less) and the execute timeout should be much higher (say 15s or more, especially if using the `createSynchronousQuery` mutation). Using the same value for all seems a bit odd.

serramatutu commented:
> especially if using the createSynchronousQuery mutation

I didn't add that to the SDK. If users run `SemanticLayerClient.query(...)`, it will always prefer to go via ADBC first. If they really want to, they can use `GraphQLClient.query(...)` directly, but that uses `createQuery` and polls `query`.

serramatutu commented:
@aiguofer I changed the PR description and title to reflect the new changes. I didn't rebase to erase the old commits to avoid erasing our discussion since Github is a bit stupid.

aiguofer commented:
This is looking great! Love all the comments explaining various things.

However, it's still missing passing the `execution_timeout` to the gql Client

aiguofer commented on dbtsl/api/graphql/client/base.py:
```diff
@@ -28,14 +29,22 @@ class BaseGraphQLClient(Generic[TTransport, TSession]):
     PROTOCOL = GraphQLProtocol
     DEFAULT_URL_FORMAT = env.DEFAULT_GRAPHQL_URL_FORMAT
+    DEFAULT_TIMEOUT = TimeoutOptions(
+        total_timeout=30,
```
30 seconds seems a little low... I wonder if we should look into some datadog logs to figure out the p75 or something?

aiguofer commented on dbtsl/api/graphql/client/base.py:
```diff
@@ -28,14 +29,22 @@ class BaseGraphQLClient(Generic[TTransport, TSession]):
+        # is the default packet retransmission window for TCP
+        # See: https://datatracker.ietf.org/doc/html/rfc2988
+        connect_timeout=4,
+        execute_timeout=10,
```
similarly... I know there's compile requests that take upwards of 30s, maybe we should also look at some DD stats to figure out an optimal value here

serramatutu commented:
> However, it's still missing passing the execution_timeout to the gql Client

I totally overlooked that lol. It's so many timeouts to pass around everywhere that I missed this one. I'll prolly add some tests too

serramatutu commented:
@aiguofer made some changes since last review:
- increased the default timeout to 60s
- passing in `execute_timeout` to the GraphQL client
- added integration tests to all this. I found some issues with the implementation, especially with `aiohttp` stuff, so I changed it to work
- added a `__post_init__` to `TimeoutOptions` to make sure everything is upper bounded by `total_timeout`

aiguofer reviewed (approved):
this is awesome, thanks!
